### PR TITLE
feat: use salts and encryption in backend to better store passwords

### LINKED ISSUE #387: Password storage: use Salts
We should use salts to improve password storing security

Implemented using bcrypt

TODO:
- [x] implement using salts
- [x] change default password in createscript used by osoc after using the salts

### PR DESCRIPTION
**Pull request by BramDevlaminck**
closes  #387 

⚠️ test this pr thoroughly!!! ⚠️

### CODE REVIEW

**Review by jay-tux — CHANGES_REQUESTED** [review]
Code looks great, and curly still works (tested on both bob & osoc; can't test trudy afaik because her email isn't stored in normalized form in the DB).

However, small remark: you use the same code in three different files. I recommend making a small function in `utility.ts` to aggregate this function (so we only have to change this function in one place, and reduce possible technical debt).

In all other aspects, amazing work 🚀

File: backend/routes/user.ts [hunk 1]
```diff
@@ -273,21 +279,36 @@ export async function userModSelf(
...
+                            loginUserId: checked.userId,
+                            accountStatus: user.account_status,
+                            password: hash,
+                        });
+                    } else {
+                        return Promise.resolve(user);
+                    }
+                })
                 .then((user) => Promise.resolve(user.person))
                 .then((person) => {
```
Comment (Mouwrice) [inline backend/routes/user.ts:313]: Not really a fan of nested `.then()` but if the backend people find this easier then using `await` that's fine by me.

**Review by Mouwrice — CHANGES_REQUESTED** [review]
Tried and tested!

**Review by Mouwrice — APPROVED** [review]

**Comment by BramDevlaminck** [comment]
> Code looks great, and curly still works (tested on both bob & osoc; can't test trudy afaik because her email isn't stored in normalized form in the DB).
> 
> However, small remark: you use the same code in three different files. I recommend making a small function in `utility.ts` to aggregate this function (so we only have to change this function in one place, and reduce possible technical debt).
> 
> In all other aspects, amazing work 🚀

what part do you see as duplicate? I only notice calculating the hash but the difference would be very minor. I would still need to call that function with 2 arguments. The only difference is that `await bcrypt.hash` would be replaced by `await ownFunc`

**Review by jay-tux — APPROVED** [review]
Dumb me 🪨

File: frontend/package.json [hunk 2]
```diff
@@ -18,12 +18,12 @@
...
   "devDependencies": {
     "@types/bcrypt": "^5.0.0",
-    "@types/node": "17.0.31",
+    "@types/node": "17.0.30",
     "@types/react": "18.0.8",
     "@types/validator": "^13.7.2",
     "cross-env": "^7.0.3",
     "eslint": "8.14.0",
-    "eslint-config-next": "12.1.6",
+    "eslint-config-next": "12.1.5",
```
Comment (Mouwrice) [inline frontend/package.json:26]: Interesting. The versions got downgraded, can this be undone?

Comment (BramDevlaminck) [reply frontend/package.json]: I've fixed this. Not sure how it happened. Probably something went wrong while merging

**Review by Mouwrice — APPROVED** [review]

**Review by nbeteram — CHANGES_REQUESTED** [review]
The password change doesn't work yet.

**Review by nbeteram — APPROVED** [review]

**Pull request merged**